sql: execute batch statements in an implicit transaction

rafiss opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/44803
relates to https://github.com/cockroachdb/cockroach/issues/76490

Release justification: high value bug fix to existing functionality 

### *: prepare tests for batch stmt txn change

This commit will make the next one easier to review.

### sql: execute batch statements in an implicit transaction

Release note (bug fix): Previously statements that arrived in a batch
during the simple query protocol would all execute in their own implicit
transactions. Now, we match the Postgres wire protocol, so all these
statements share the same implicit transaction. If a BEGIN is included
in a statement batch, then the existing implicit transaction is
upgraded to an explicit transaction.

### sql: add session var for old implicit txn behavior

Release note (sql change): The enable_implicit_transaction_for_batch_statements
session variable was added. It defaults to true. When it is true,
multiple statements in a single query (a.k.a. a "batch statement") will
all be run in the same implicit transaction, which matches the Postgres
wire protocol. This setting is provided for users who want to preserve
the behavior of CockroachDB versions v21.2 and earlier.

rafiss committed:
*: prepare tests for batch stmt txn change

This commit will make the next one easier to review.

Release justification: test only change

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76834)
<!-- Reviewable:end -->

otan reviewed (commented):
change by itself lgtm, but i think this may cause some tests to fail as they use multiple schema changes in the same cmd. the test failures seem suspicious as a result.

otan commented:
also, does this break if someone does something like `./cockroach sql < lots_of_schema_statements.sql`

rafiss commented:
> also, does this break if someone does something like ./cockroach sql < lots_of_schema_statements.sql

well depends what you mean by "break." :P

but yes,  pretty sure that will cause all the statements to be run inside of the same implicit txn. but you could also say that the "broken" part is the fact that schema changes have odd behavior inside of a transaction. for that case, we could potentially make `cockroach sql` explicitly send each statement in a separate SimpleQuery. i'll look into it

this change is looking like it will have a bunch of test fall out, so i'm thinking of waiting until after stability period to get it in, even though it's a bug fix. unless you have any thoughts?

knz commented:
> pretty sure that will cause all the statements to be run inside of the same implicit txn. 

Have you checked? IIRC the shell will read the file line by line (breaking the input at newline characters) and issue them in separate SQL queries.

knz commented:
> this change is looking like it will have a bunch of test fall out,

I've looked at the failing tests. They seem to be failing for innocuous reasons and should be straightforward to adjust. I personally don't see this as a risk during stability period.

rafiss commented:
> Have you checked? IIRC the shell will read the file line by line (breaking the input at newline characters) and issue them in separate SQL queries.

Just checked now, and it looks like it's sent all at once.

Before my change:
```
❯ ./cockroach-short sql --insecure -e "create table a(a int); insert into a values( 1); select 1/0; select 2;";
ERROR: division by zero
SQLSTATE: 22012
Failed running "sql"

❯ ./cockroach-short sql --insecure -e "select * from a;";
  a
-----
  1
(1 row)


Time: 1ms
```

After my change:
```
❯ ./cockroach-short sql --insecure -e "create table a(a int); insert into a values( 1); select 1/0; select 2;";
ERROR: division by zero
SQLSTATE: 22012
Failed running "sql"
❯ ./cockroach-short sql --insecure -e "select * from a;";
ERROR: relation "a" does not exist
SQLSTATE: 42P01
Failed running "sql"
```

---

> I've looked at the failing tests. They seem to be failing for innocuous reasons and should be straightforward to adjust. I personally don't see this as a risk during stability period.

Thanks, it's helpful to hear. I also would prefer to merge during the stability period, but I second-guessed myself after wondering if I was too over-eager.

rafiss commented:
Ah I see, when using `-f` with a file argument, the statements are all run separately: https://github.com/cockroachdb/cockroach/blob/9f350c60fc5b2fc28f059f1c7a9f9b81e6407b2d/pkg/cli/clisqlshell/sql.go#L1453-L1454

But sith `-e` and a string argument, the statements are sent all in one batch: https://github.com/cockroachdb/cockroach/blob/9f350c60fc5b2fc28f059f1c7a9f9b81e6407b2d/pkg/cli/clisqlshell/sql.go#L1469-L1470

otan commented:
my concern with the implicit transaction approach is previous schema changes are now txn-inal and from the tests that broke we might have problems with that now...

knz commented:
for `cockroach sql`:
- `-f` : file is read line by line
- reading from a file redirected on stdin: line by line
- `-e`: one batch per use of `-e` (you can use `-e` multiple times)

For the unit tests: yes, some of our unit tests were using multiple statements separated with semicolons before. Arguably this was a misdesign, born out of convenience. I don't think it hurts much to split these batches into multiple exec/query calls.

otan commented on pkg/testutils/jobutils/jobs_verification.go:
```diff
@@ -154,6 +154,8 @@ func verifySystemJob(
 	sort.Sort(expected.DescriptorIDs)
 	expected.Details = nil
 	if e, a := expected, actual; !reflect.DeepEqual(e, a) {
+		fmt.Printf("%+v\n", expected)
```
nit: delete, or use `t.Logf`

otan reviewed (commented):
changes LGTM, i wonder if we should add an extra appendix on the release note that certain schema change ops can no longer be sent "as a single batch" because it *might* break as we may not support certain schema changes in the same txn.

i still have a hangup from above though. it's going to look weird to a user copying  a big schema blob in that their schema change fails in some strange way because we don't support transactional schema changes yet.

ajwerner commented:
Is there any way we can gate this behavior?

rafiss commented:
Yes I can add a session variable to control this here: https://github.com/cockroachdb/cockroach/pull/76834/files#diff-30926eec6a83d84431b319a14fc9081df5096743cff7914a77f3bdbfa5a251c7

otan commented:
i think the session variable would be gr8, just in case :)

rafiss committed:
sql: execute batch statements in an implicit transaction

Release justification: high value bug fix to existing functionality.

Release note (bug fix): Previously statements that arrived in a batch
during the simple query protocol would all execute in their own implicit
transactions. Now, we match the Postgres wire protocol, so all these
statements share the same implicit transaction. If a BEGIN is included
in a statement batch, then the existing implicit transaction is
upgraded to an explicit transaction.

rafiss committed:
sql: add session var for old implicit txn behavior

Release justification: low risk new setting.

Release note (sql change): The enable_implicit_transaction_for_batch_statements
session variable was added. It defaults to true. When it is true,
multiple statements in a single query (a.k.a. a "batch statement") will
all be run in the same implicit transaction, which matches the Postgres
wire protocol. This setting is provided for users who want to preserve
the behavior of CockroachDB versions v21.2 and earlier.

otan commented on pkg/sql/conn_executor.go:
```diff
@@ -1846,7 +1846,10 @@ func (ex *connExecutor) execCmd() error {
-			canAutoCommit := ex.implicitTxn() && tcmd.LastInBatch
+			// The behavior is configurable, in case users want to preserve the
+			// behavior from v21.2 and earlier.
+			implicitTxnForBatch := ex.sessionData().EnableImplicitTransactionForBatchStatements
```
wondering: if someone does `ALTER DATABASE d SET enable_implicit_transaction_for_batch_statements = false` and rolls back, does their database not load because the session variable cannot be set? i wonder if people would want this behaviour with mixed versions.

also, do we want to metamorphic test this or nah?

rafiss replied:
> does their database not load because the session variable cannot be set

do you mean the setting that happens during session init? we should be good there. a warning is logged and it continues: https://github.com/cockroachdb/cockroach/blob/29797de1825eb1a12c357f899ef2308a60ebf8e0/pkg/sql/pgwire/auth.go#L189-L191

hm yeah i suppose metamorphic won't hurt

rafiss replied:
actually, a metamorphic setting might hurt, since it would make traces non-deterministic. i guess we can try then see if tests are affected

otan replied:
if the tests are tricky im happy to let it go. but i think this looks fine.

my only other concern here is you need to migrate at least one node to move before setting the default var. but then again, you could argue that users should have these implicit statements split out like you have in the tests in v21.2 before migrating to v22.1 anyway, so a backwards compatible pathway is available.

rafiss commented:
> if the tests are tricky im happy to let it go. but i think this looks fine.

yeah i'm going to un-metamorphic it. it's like whack-a-mole to find which tests are affected. in retrospect, i think metamorphic constants should not influence any user-facing logical behavior.

rafiss commented:
bors r=otan

rafiss commented:
bors r=otan

rafiss commented:
bors r=otan

rafiss commented:
bors r-

rafiss commented:
bors r=otan
